tsec-http4s/src/main/scala/tsec/authentication/JWKSAuthenticator.scala

```diff
@@ -0,0 +1,185 @@
+package tsec.authentication
+
+import java.math.BigInteger
+import java.time.{Instant, LocalDateTime}
+import java.util.Base64
+
+import cats.data.OptionT
+import cats.effect.{Effect, Sync}
+import cats.syntax.all._
+import io.circe.Decoder
+import io.circe.parser.decode
+import org.http4s.headers.Authorization
+import org.http4s.{AuthScheme, Credentials, EntityDecoder, Request, Response, Uri}
+import tsec.common.SecureRandomId
+import tsec.jws.signature.{JWSSigCV, JWTSig}
+import tsec.signature.jca.{JCASigTag, SigPublicKey}
+
+import scala.concurrent.duration.FiniteDuration
+
+final case class AugmentedJWKS[A, I](
+    id: SecureRandomId,
+    jwt: JWTSig[A],
+    identity: I,
+    expiry: Instant,
+    lastTouched: Option[Instant]
+)
+
+final case class JWK(kid: String, alg: String, kty: String, use: String, n: String, e: String)
+
+final case class JWKS(keys: List[JWK])
+
+class KeyRegistry[F[_], A: JCASigTag](uri: Uri, minFetchDelay: FiniteDuration)(implicit E: Effect[F]) {
+
+  import io.circe.generic.auto._
+  import org.http4s.circe._
+  import org.http4s.client.blaze._
+  import java.security.KeyFactory
+  import java.security.spec.RSAPublicKeySpec
+
+  implicit val jwksDecoder: EntityDecoder[F, JWKS] = jsonOf[F, JWKS]
+
+  private var keys = Map[String, SigPublicKey[A]]()
```

Comment:
This is making me feel like this could be accessed in a multithreaded scenario.

Use a `ref` or an `atomicReference` pls.